File: src/scanner.js
import fs from 'fs';
import path from 'path';

const DEFAULT_IGNORES = new Set([
  'node_modules',
  '.git',
  '.svn',
  '.hg',
  'venv',
  '.venv',
  'env',
  'dist',
  'build',
  'out',
  '.next',
  '.nuxt',
  '.turbo',
  '.cache',
  'coverage',
  '.pytest_cache',
  '__pycache__',
  'vendor',
  '.idea',
  '.vscode',
  'target',
  'tmp',
  'web',
  'site',
  'docs',
  'tests',
  'test',
  '__tests__',
  'fixtures'
]);

const VALID_EXTENSIONS = new Set([
  '.js', '.mjs', '.cjs', '.jsx',
  '.ts', '.mts', '.cts', '.tsx',
  '.py', '.pyw',
  '.go',
  '.rs',
  '.php',
  '.sh', '.bash', '.zsh'
]);

const SPECIAL_FILENAMES = new Set([
  'Dockerfile',
  'docker-compose.yml',
  'docker-compose.yaml'
]);

// Scoped pattern definitions by language/file type
const JS_PATTERNS = [
  { regex: /process\.env\.([A-Z0-9_]+)/g, extract: m => m[1] },
  { regex: /process\.env\[['"]([A-Z0-9_]+)['"]\]/g, extract: m => m[1] },
  { regex: /import\.meta\.env\.([A-Z0-9_]+)/g, extract: m => m[1] }
];

const PYTHON_PATTERNS = [
  { regex: /os\.environ\.get\(\s*['"]([A-Z0-9_]+)['"]/g, extract: m => m[1] },
  { regex: /os\.getenv\(\s*['"]([A-Z0-9_]+)['"]/g, extract: m => m[1] },
  { regex: /os\.environ\[\s*['"]([A-Z0-9_]+)['"]\s*\]/g, extract: m => m[1] }
];

const GO_PATTERNS = [
  { regex: /os\.(?:Getenv|LookupEnv)\(\s*"([A-Z0-9_]+)"\s*\)/g, extract: m => m[1] }
];

const PHP_PATTERNS = [
  { regex: /(?<![\.\$])\bgetenv\(\s*['"]([A-Z0-9_]+)['"]\s*\)/g, extract: m => m[1] },
  { regex: /\$_(?:ENV|SERVER)\[\s*['"]([A-Z0-9_]+)['"]\s*\]/g, extract: m => m[1] }
];

const SHELL_PATTERNS = [
  { regex: /\$\{([A-Z0-9_]{3,})\}/g, extract: m => m[1] },
  { regex: /^ENV\s+([A-Z0-9_]+)/gm, extract: m => m[1] }
];

const SYSTEM_IGNORES = new Set([
  'NODE_ENV',
  'PATH',
  'HOME',
  'USER',
  'SHELL',
  'PWD',
  'TERM',
  'LANG',
  'TMPDIR',
  'SHLVL',
  'CI',
  'GITHUB_ACTIONS',
  'VERCEL',
  'NETLIFY'
]);

function getPatternsForFile(filePath) {
  const ext = path.extname(filePath).toLowerCase();
  const basename = path.basename(filePath);

  if (basename === 'Dockerfile') {
    return SHELL_PATTERNS;
  }
  if (['.js', '.mjs', '.cjs', '.jsx', '.ts', '.mts', '.cts', '.tsx'].includes(ext)) {
    return JS_PATTERNS;
  }
  if (['.py', '.pyw'].includes(ext)) {
    return PYTHON_PATTERNS;
  }
  if (ext === '.go') {
    return GO_PATTERNS;
  }
  if (ext === '.php') {
    return PHP_PATTERNS;
  }
  if (['.sh', '.bash', '.zsh'].includes(ext)) {
    return SHELL_PATTERNS;
  }
  return [];
}

/**
 * Recursively find all eligible files in a directory.
 */
export function findFiles(dir, rootDir = dir, customIgnores = []) {
  const ignoreSet = new Set([...DEFAULT_IGNORES, ...customIgnores]);
  const results = [];

  let entries;
  try {
    entries = fs.readdirSync(dir, { withFileTypes: true });
  } catch {
    return results;
  }

  for (const entry of entries) {
    const fullPath = path.join(dir, entry.name);

    if (entry.isDirectory()) {
      if (!ignoreSet.has(entry.name) && !entry.name.startsWith('.')) {
        results.push(...findFiles(fullPath, rootDir, customIgnores));
      }
    } else if (entry.isFile()) {
      const ext = path.extname(entry.name).toLowerCase();
      if (VALID_EXTENSIONS.has(ext) || SPECIAL_FILENAMES.has(entry.name)) {
        if (entry.name.endsWith('.min.js') || entry.name.endsWith('.lock') || entry.name === 'package-lock.json') {
          continue;
        }
        results.push(fullPath);
      }
    }
  }

  return results;
}

/**
 * Scan a single file for env variable usages.
 */
export function scanFile(filePath, rootDir = process.cwd()) {
  const matches = [];
  const patterns = getPatternsForFile(filePath);
  if (patterns.length === 0) return matches;

  let content;
  try {
    content = fs.readFileSync(filePath, 'utf8');
  } catch {
    return matches;
  }

  const lines = content.split(/\r?\n/);
  const relPath = path.relative(rootDir, filePath);

  lines.forEach((lineText, idx) => {
    const lineNum = idx + 1;
    const trimmed = lineText.trim();
    if (trimmed.startsWith('//') || trimmed.startsWith('#') || trimmed.startsWith('*')) {
      return;
    }

    for (const { regex, extract } of patterns) {
      regex.lastIndex = 0;
      let match;
      while ((match = regex.exec(lineText)) !== null) {
        const varName = extract(match);
        if (varName && !SYSTEM_IGNORES.has(varName) && /^[A-Z][A-Z0-9_]*$/.test(varName)) {
          matches.push({
            name: varName,
            file: relPath,
            line: lineNum,
            snippet: lineText.trim()
          });
        }
      }
    }
  });

  return matches;
}

/**
 * Read ignore patterns from .gitignore if present.
 */
export function loadGitignorePatterns(rootDir) {
  const gitignorePath = path.join(rootDir, '.gitignore');
  if (!fs.existsSync(gitignorePath)) return [];
  try {
    const content = fs.readFileSync(gitignorePath, 'utf8');
    return content
      .split(/\r?\n/)
      .map(line => line.trim())
      .filter(line => line && !line.startsWith('#'))
      .map(line => line.replace(/^\//, '').replace(/\/$/, ''));
  } catch {
    return [];
  }
}

/**
 * Scan entire codebase starting at rootDir.
 */
export function scanCodebase(rootDir = process.cwd(), customIgnores = []) {
  const gitignorePatterns = loadGitignorePatterns(rootDir);
  const files = findFiles(rootDir, rootDir, [...gitignorePatterns, ...customIgnores]);
  const varMap = new Map();

  for (const file of files) {
    const hits = scanFile(file, rootDir);
    for (const hit of hits) {
      if (!varMap.has(hit.name)) {
        varMap.set(hit.name, {
          name: hit.name,
          occurrences: []
        });
      }
      varMap.get(hit.name).occurrences.push({
        file: hit.file,
        line: hit.line,
        snippet: hit.snippet
      });
    }
  }

  return varMap;
}

File: dotvet/scanner.py
import os
import re
from pathlib import Path
from typing import Dict, List, Any

DEFAULT_IGNORES = {
    "node_modules",
    ".git",
    ".svn",
    ".hg",
    "venv",
    ".venv",
    "env",
    "dist",
    "build",
    "out",
    ".next",
    ".nuxt",
    ".turbo",
    ".cache",
    "coverage",
    ".pytest_cache",
    "__pycache__",
    "vendor",
    ".idea",
    ".vscode",
    "target",
    "tmp",
}

VALID_EXTENSIONS = {
    ".js", ".mjs", ".cjs", ".jsx",
    ".ts", ".mts", ".cts", ".tsx",
    ".py", ".pyw",
    ".go",
    ".rs",
    ".rb",
    ".php",
    ".sh", ".bash", ".zsh",
    ".yaml", ".yml",
    ".json",
}

SPECIAL_FILENAMES = {
    "Dockerfile",
    "docker-compose.yml",
    "docker-compose.yaml",
    ".env.example",
    ".env.sample",
    ".env.template",
}

PATTERNS = [
    # JS/TS: process.env.FOO or process.env['FOO']
    re.compile(r"process\.env\.([A-Z0-9_]+)"),
    re.compile(r"process\.env\[['\"]([A-Z0-9_]+)['\"]\]"),
    # Vite / ESM: import.meta.env.VITE_FOO
    re.compile(r"import\.meta\.env\.([A-Z0-9_]+)"),
    # Python: os.environ.get('FOO'), os.getenv('FOO'), os.environ['FOO']
    re.compile(r"os\.environ\.get\(\s*['\"]([A-Z0-9_]+)['\"]"),
    re.compile(r"os\.getenv\(\s*['\"]([A-Z0-9_]+)['\"]"),
    re.compile(r"os\.environ\[\s*['\"]([A-Z0-9_]+)['\"]\s*\]"),
    # Go: os.Getenv("FOO"), os.LookupEnv("FOO")
    re.compile(r'os\.(?:Getenv|LookupEnv)\(\s*"([A-Z0-9_]+)"\s*\)'),
    # Shell / Docker: ${VAR_NAME} or ENV VAR_NAME=...
    re.compile(r"\$\{([A-Z0-9_]{3,})\}"),
    re.compile(r"^ENV\s+([A-Z0-9_]+)", re.MULTILINE),
]

SYSTEM_IGNORES = {
    "NODE_ENV",
    "PATH",
    "HOME",
    "USER",
    "SHELL",
    "PWD",
    "TERM",
    "LANG",
    "TMPDIR",
    "SHLVL",
    "CI",
}


def find_files(root_dir: str, custom_ignores: List[str] = None) -> List[str]:
    """Recursively find all eligible files in root_dir."""
    ignore_set = DEFAULT_IGNORES | set(custom_ignores or [])
    matched_files = []
    root_path = Path(root_dir).resolve()

    for dirpath, dirnames, filenames in os.walk(root_path):
        # Filter directories in-place to avoid descending into ignored folders
        dirnames[:] = [
            d for d in dirnames
            if d not in ignore_set and not d.startswith(".")
        ]

        for fname in filenames:
            ext = Path(fname).suffix.lower()
            if ext in VALID_EXTENSIONS or fname in SPECIAL_FILENAMES:
                if fname.endswith(".min.js") or fname.endswith(".lock") or fname == "package-lock.json":
                    continue
                matched_files.append(os.path.join(dirpath, fname))

    return matched_files


def scan_file(file_path: str, root_dir: str) -> List[Dict[str, Any]]:
    """Scan a single file for environment variable references."""
    matches = []
    try:
        with open(file_path, "r", encoding="utf-8", errors="ignore") as f:
            content = f.read()
    except Exception:
        return matches

    lines = content.splitlines()
    rel_path = os.path.relpath(file_path, root_dir)

    for idx, line_text in enumerate(lines):
        line_num = idx + 1
        trimmed = line_text.strip()
        if trimmed.startswith("//") or trimmed.startswith("#") or trimmed.startswith("*"):
            continue

        for pattern in PATTERNS:
            for match in pattern.finditer(line_text):
                var_name = match.group(1)
                if var_name and var_name not in SYSTEM_IGNORES and re.match(r"^[A-Z][A-Z0-9_]*$", var_name):
                    matches.append({
                        "name": var_name,
                        "file": rel_path,
                        "line": line_num,
                        "snippet": trimmed,
                    })

    return matches


def scan_codebase(root_dir: str = ".", custom_ignores: List[str] = None) -> Dict[str, Dict[str, Any]]:
    """Scan entire codebase and return map of var_name -> metadata."""
    files = find_files(root_dir, custom_ignores)
    var_map: Dict[str, Dict[str, Any]] = {}

    for fpath in files:
        hits = scan_file(fpath, root_dir)
        for hit in hits:
            name = hit["name"]
            if name not in var_map:
                var_map[name] = {
                    "name": name,
                    "occurrences": [],
                }
            var_map[name]["occurrences"].append({
                "file": hit["file"],
                "line": hit["line"],
                "snippet": hit["snippet"],
            })

    return var_map
